### PR TITLE
HAWQ-1439. tolerate system time being changed to earlier point when checking resource context timeout

### PR DESCRIPTION
**Pull request by jiny2**

### CODE REVIEW

**Comment by jiny2** [comment]
@paul-guo- @liming01  please review

**Comment by paul-guo-** [comment]
This is a workaround. A better solution seems to use the relative time based on H/W tick. I'd test to use clock_gettime() to replace gettimeofday() if clock_gettime() exists.

**Comment by liming01** [comment]
+1. LGTM.

**Comment by jiny2** [comment]
I think as we discussed, we will do a simple fix firstly for this problem. Changing gettime_microsec() may have side effect which is not clear yet.

**Comment by jiny2** [comment]
Seems clock_gettime() does not work for C99. 

network_utils.c:52:11: warning: implicit declaration of function 'clock_gettime' is invalid in C99 [-Wimplicit-function-declaration]
        status = clock_gettime(CLOCK_MONOTONIC, &ts);
                 ^
network_utils.c:52:25: error: use of undeclared identifier 'CLOCK_MONOTONIC'
        status = clock_gettime(CLOCK_MONOTONIC, &ts);
                               ^

**Comment by jiny2** [comment]
Applied @paul-guo- 's comment to change the time getting implementation

File: src/backend/resourcemanager/utils/network_utils.c [hunk 1]
```diff
@@ -41,9 +42,23 @@ static void cleanupSocketConnectionPool(int code, Datum arg);
...
+	newTime.tv_sec = ts.tv_sec;
+	newTime.tv_usec = ts.tv_nsec / 1000;
+#endif
+
+	if (status != 0)
+	{
+		gettimeofday(&newTime, NULL);
+	}
+	t = ((uint64_t)newTime.tv_sec) * USECS_PER_SECOND + newTime.tv_usec;
+    return t;
```
Comment (paul-guo-) [inline src/backend/resourcemanager/utils/network_utils.c]: spaces -> tab.

File: src/backend/resourcemanager/utils/network_utils.c [hunk 2]
```diff
@@ -29,6 +29,7 @@
 #include <stdlib.h>
 #include <unistd.h>
 #include <sys/types.h>
+#include <time.h>
```
Comment (paul-guo-) [inline src/backend/resourcemanager/utils/network_utils.c]: You might add the following header file to suppress the warning about missing definition of on_proc_exit().

#include "storage/ipc.h"

**Comment by jiny2** [comment]
@paul-guo-  please double check the latest changes

**Pull request closed without merging**